app/lib/components/scrollbar/Scrollbar.tsx

```diff
@@ -0,0 +1,170 @@
+
+    this.thumb!.style.left = `${this.thumbX}px`;
+    this.thumb!.style.width = `${this.thumbWidth}px`;
+
+    this.track.setAttribute(
+      "aria-valuenow",
+      String(
+        Math.round(
+          this.thumbX === 0 ? 0 : (100 * this.thumbX) / (trackClientWidth - this.thumbWidth)
+        )
+      )
+    );
+  }
+
+  private isScrolling = false;
+  // Single scroll event here avoids creating too many objects.
+  private scrollEvent = { delta: 0, animate: false };
+  private minMouseX = 0;
+  private maxMouseX = 0;
+  private mouseX = 0;
+  onMouseDown = (e: MouseEvent) => {
+    if (e.buttons & 1) {
+      this.isScrolling = true;
+      const thumb = this.thumb!.getBoundingClientRect();
+      const track = this.track!.getBoundingClientRect();
+      this.minMouseX = e.clientX - (thumb.x - track.x);
+      this.maxMouseX = e.clientX + (track.x + track.width - (thumb.x + thumb.width));
+    }
+    this.updateMouse(e);
+  };
+  onMouseMove = (e: MouseEvent) => {
+    if (!this.isScrolling) return;
+    if (!(e.buttons & 1)) {
+      this.isScrolling = false;
+      return;
+    }
+
+    const lastX = this.mouseX;
+    this.updateMouse(e);
+    const deltaX = this.mouseX - lastX;
+
+    if (deltaX !== 0 && this.track!.clientWidth - this.thumbWidth !== 0) {
+      this.scrollEvent.delta =
+        (deltaX / (this.track!.clientWidth - this.thumbWidth)) * this.scrollLeftMax;
+      this.scrollEvent.animate = false;
+      this.events.dispatch("scroll", this.scrollEvent);
+    }
+  };
+  onMouseUp = (e: MouseEvent) => {
+    this.updateMouse(e);
+    this.isScrolling = false;
+  };
+  updateMouse = (e: MouseEvent) => {
+    this.mouseX = Math.min(this.maxMouseX, Math.max(this.minMouseX, e.clientX));
+  };
+
+  private isMouseInside = false;
+  onMouseEnterScrollingElement = (e: MouseEvent) => {
+    this.isMouseInside = true;
+  };
+  onMouseOutScrollingElement = (e: MouseEvent) => {
+    this.isMouseInside = false;
+  };
+  onWheelScrollingElement = (e: WheelEvent) => {
+    if (e.shiftKey) {
+      e.preventDefault();
+      this.scrollEvent.delta = e.deltaY;
+      this.scrollEvent.animate = true;
+      this.events.dispatch("scroll", this.scrollEvent);
+    }
+  };
+  onKeyDown = (e: KeyboardEvent) => {
+    if (!this.isMouseInside || e.ctrlKey || e.shiftKey) return;
+
+    if (e.which === 39 || e.which === 37) {
+      e.preventDefault();
+      const dir = e.which - 38;
+      this.scrollEvent.delta = 40 * dir;
+      this.scrollEvent.animate = true;
+      this.events.dispatch("scroll", this.scrollEvent);
+    }
+  };
+
+  dispose() {
+    this.disposer.dispose();
+  }
+}
+
+export type HorizontalScrollbarProps = {
+  setController: (controller: HorizontalScrollbarController) => void;
+};
+
+export function HorizontalScrollbar({ setController }: HorizontalScrollbarProps) {
+  const trackRef = React.useRef<HTMLDivElement>(null);
+
+  React.useEffect(() => {
```

Comment:
this feels like an initializer if it were a class.

what's the benefit of separating out the controller from this little bit of logic & render? can we just make this one component and simplify things?

Reply:
I lean towards making everything a function these days since they allow you to use hooks, and when the function starts getting too large (too many `useCallback` etc.) I've started using this pattern of moving the event listeners and DOM manipulation logic into a separate class.

I place a lot of value on the ability to use hooks since they are a great way to add functionality to a component without worrying about having to clean it up manually, and they also enable certain features like consuming multiple contexts and simpler "rerun if this value has changed" logic (via deps arrays).

There's another philosophy I've seen -- "start with a class component, then migrate to function if you need to use hooks." But with this approach I often wind up needing to use hooks and then have to convert the class to a function component anyway, which can be tedious and error prone.